### DISCUSSION TITLE
Is branch Malloc with memory leak function (nativemem) correct ?

### DISCUSSION THREAD

**Discussion by ksnovich** [async-profiler/async-profiler#649] [Q&A] [answered]
Is branch "Malloc" with memory leak function (nativemem) correct ?

I don't understand what do I do wrong..

I'm going to check my Java application for memory leak with Andrey's new function memore leak (nativemem) where use malloc&free.
I downloaded https://profiler.tools/builds/async-profiler-2.1-malloc-linux-x64.tar.gz

Start my app like this:
LD_PRELOAD=/tmp/async-profiler-2.1-malloc-linux-x64/build/libasyncProfiler.so $JAVA_HOME/bin/java -agentpath:/tmp/async-profiler-2.1-malloc-linux-x64/build/libasyncProfiler.so=start,event=nativemem,file=malloc.jfr .....

There aren't any problem.

Java application is Spring Boot app.

Waited about 60 seconds.

Stopped profiling by execute ./profiler.sh stop 25936 > malloc.out

Stopped application.

After that I converted malloc.jfr to html by converter:
	java -cp ./converter.jar MallocReport malloc.jfr malloc.html

And I was really confused because in the FlameGraph malloc.html were a lot of stacks different code.

Also, size of malloc.out is small 165Kb but size of malloc.jfr is really big ~ 182Mb.

Unfortunately, I couldn't attach my files.

I started compare malloc.html(flamegraph) and malloc.out(txt) files.

There are a lot of different Spring boot code in the FlameGrap file. For example Hiberante, Apache and etc.

But, in the malloc.out(txt) file there are a few code. 
In general, there are C2Compiler stacks. 

And between C2Compiler codes I can see code like:
--- 51642368 bytes (0.38%), 0 samples
  [ 0] Profiler::getNativeTrace(Engine*, void*, ASGCT_CallFrame*, int)
  [ 1] Profiler::recordSample(void*, unsigned long long, int, Event*)
  [ 2] malloc
  [ 3] updatewindow
  [ 4] inflate
  [ 5] Java_java_util_zip_Inflater_inflateBytesBytes
  [ 6] java.util.zip.Inflater.inflateBytesBytes
  [ 7] java.util.zip.Inflater.inflate
  [ 8] java.util.zip.InflaterInputStream.read
  [ 9] org.springframework.boot.loader.jar.ZipInflaterInputStream.read
  ...

....
  [ 0] Profiler::getNativeTrace(Engine*, void*, ASGCT_CallFrame*, int)
  [ 1] Profiler::recordSample(void*, unsigned long long, int, Event*)
  [ 2] malloc
  [ 3] updatewindow
  [ 4] inflate
  [ 5] Java_java_util_zip_Inflater_inflateBytesBytes
  [ 6] java.util.zip.Inflater.inflateBytesBytes
  [ 7] java.util.zip.Inflater.inflate
  [ 8] java.util.zip.InflaterInputStream.read
  [ 9] org.springframework.boot.loader.jar.ZipInflaterInputStream.read
  [10] jdk.internal.loader.Resource.getBytes
  [11] java.net.URLClassLoader.defineClass
  [12] java.net.URLClassLoader$1.run
  [13] java.net.URLClassLoader$1.run
  [14] java.security.AccessController.doPrivileged
  [15] java.net.URLClassLoader.findClass
  [16] java.lang.ClassLoader.loadClass
  [17] org.springframework.boot.loader.LaunchedURLClassLoader.loadClass
  [18] java.lang.ClassLoader.loadClass
  [19] java.lang.Class.forName0
  [20] java.lang.Class.forName
  [21] org.springframework.util.ClassUtils.forName
  [22] org.springframework.boot.liquibase.SpringPackageScanClassResolver.loadClass
  ...

I confused, why I couldn't find Hibernate code in the malloc.out file opposite malloc.html file ?
Is Inflater really memory leak ?

Thanks for watching my long description :)

p.s. Also I tried to use myself build from "malloc" branch  but I couldn't execute it.
When I start JVM with new lib so I get mistakes:
/usr/lib64/jvm/jdk-11-0-13/bin/java: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.20' not found (required by xxxxxxxxxx/async-malloc-ksn/build/libasyncProfiler.so)
/usr/lib64/jvm/jdk-11-0-13/bin/java: /lib64/libstdc++.so.6: version `CXXABI_1.3.8' not found (required by xxxxxxxxxx/async-malloc-ksn/build/libasyncProfiler.so)
/usr/lib64/jvm/jdk-11-0-13/bin/java: /lib64/libstdc++.so.6: version `CXXABI_1.3.9' not found (required by xxxxxxxxxx/async-malloc-ksn/build/libasyncProfiler.so)
/usr/lib64/jvm/jdk-11-0-13/bin/java: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.21' not found (required by xxxxxxxxxx/async-malloc-ksn/build/libasyncProfiler.so)

There is RHEL 7.9 with kernel 3.10.0-1160.25.1.el7.x86_64  and myself Async-profiler build from "master" branch works without problem.

**Comment by apangin** [comment, marked as answer]
Since there are many different points mixed in the post, it's hard to extract the actual question.

No profiler can tell you exactly where the leak is, because it's a developer who decides whether a certain allocation pattern can be considered a leak in their particular case or not.

In `nativemem` mode, async-profiler records where the native memory is allocated and released. The raw jfr output in this mode can be indeed very big, since it includes all malloc/free calls. To reduce the amount of samples, increase `interval` parameter.

`MallocReport` then parses the raw output and shows you allocations that were not freed by the end of profiling session. Take a careful look at those allocations, when you know that memory should have been certainly freed - this is a likely sign of leak. `malloc.out(txt)` is probably not what you are looking for.

I have recently reworked `nativemem` mode - now it does not require LD_PRELOAD, works both on Linux and macOS and supports different allocators (like jemalloc and tcmalloc). This version has not been open sourced though.

**Comment by evanzhang87** [comment]
I have the same question with GLIBCXX and CXXABI, how to fix?
